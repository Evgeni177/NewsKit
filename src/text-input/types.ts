import {SizingKeys, SpacePresetKeys} from '../theme';

export enum TextInputSize {
  Small = 'small',
  Medium = 'medium',
  Large = 'large',
}

export interface TextInputProps
  extends Omit<React.InputHTMLAttributes<HTMLInputElement>, 'size'> {
  size?: TextInputSize;
  disabled?: boolean;
  label: string;
  hideLabel?: boolean;
  assistiveText?: string;
  ariaLabel?: string;
  overrides?: {
    width?: SizingKeys | string;
    input?: {
      stylePreset?: string;
      paddingPreset?: string;
      minHeight?: SizingKeys | string;
      typographyPreset?: string;
      spaceStack?: SpacePresetKeys;
      spaceInline?: SpacePresetKeys;
    };
    label?: {
      stylePreset?: string;
      typographyPreset?: string;
      spaceStack?: SpacePresetKeys;
      spaceInline?: SpacePresetKeys;
    };
    assistiveText?: {
      stylePreset?: string;
      typographyPreset?: string;
    };
  };
}
