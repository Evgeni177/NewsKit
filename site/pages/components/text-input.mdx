import Layout from '../../components/layout';
import Playground from '../../components/playground';
import {Prop} from '../../components/prop';
import {TextBoxSwatch} from '../../components/theming-values/colors';
import {LegacyBlock} from '../../components/legacy-block';
import {CodeFromDefaultPresets} from '../../components/code';
import {TextInput, TextInputSize} from 'newskit';

export default Layout;

# Text Input

## Overview

The Text Input component consists of a label, input and assistive text.

## Usage

Text Input allows users to enter text into a UI. They typically appear in forms and dialogs.


## Props

Text Input has a number of props to facilitate a variety of uses:

<LegacyBlock marginTop="sizing040">
  <Prop name="size" type="TextInputSize" enum={TextInputSize}
    default={TextInputSize.Medium}>
    Optional size of the Text Input: Small, Medium or Large.
  </Prop>
  <Prop name="label" type="string" required>
    Label text is used to inform users as to what information is requested for a text input. Every text input field should have a label.
  </Prop>
  <Prop name="hideLabel" type="boolean" default={false}>
    A prop can be set on the Text Input component to visually hide the label. 
    The label should remain readable by screen readers by setting the arial label attribute on the input to the value of the label text.
  </Prop>
  <Prop name="assistiveText" type="string">
    Assistive text gives context about a fieldâ€™s input, such as how the input will be used. It will be displayed beneath the text input. 
  </Prop>
  <Prop name="disabled" type="boolean" default={false}>
    If true, the Text Input will render the disabled state of the style preset.
  </Prop>
  <Prop name="readOnly" type="boolean" default={false}>
    If true, the Text Input will render the read only state of the style preset.
  </Prop>
  <Prop name="ariaLabel" type="string">
    The aria-label attribute for the text input. 
    As stated above, if the label is hidden, the label text will be placed into the aria label attribute of the input
    However, this can be overridden if the consumer sets aria-label prop.
  </Prop>
  <Prop name="overrides" type="object">
    If provided, overrides the respective presets for the component and provided elements.
    <Prop name="width" type="SizingKeys | string">
      If provided, this sets a fixed width to the text input. This can be a sizing token
      from the theme, or any CSS length value, e.g. 100% for a full-width element.
    </Prop>
    <Prop name="input" type="object">
      <Prop name="placeholder" type="string" default="Placeholder">
        If provided, this overrides the placeholder text applied to the text input.
      </Prop>
      <Prop name="stylePreset" type="string" default="textInput">
        If provided, this overrides the style preset applied to the text input.
      </Prop>
      <Prop name="padding" type="PaddingPresetKeys">
        If provided, this overrides the padding of the Text Input. Otherwise,
        spaceInset010Squish or spaceInset030 is used depending on the `size`
        prop.
      </Prop>
      <Prop name="typographyPreset" type="string">
        If provided, this overrides the type preset applied to the text input. Otherwise,
        body020 or body030 is used depending on the `size` prop.
      </Prop>
      <Prop name="spaceStack" type="SpacingPresetKeys">
        This is a spacing preset token which is applied as a margin on the bottom of
        the Block. Use this to space stacking content down the page.
      </Prop>
      <Prop name="minHeight" type="SizingKeys | string">
        If provided, this sets a minimum height to the Text Input. This can be a sizing
        token from the theme, or any CSS length value. By default, sizing060,
        sizing080 or sizing090 is used depending on the `size` prop.
      </Prop>
    </Prop>
    <Prop name="inputLabel" type="object">
      <Prop name="stylePreset" type="string" default="textinputLabel">
        If provided, this overrides the style preset applied to the text input label.
      </Prop>
      <Prop name="typographyPreset" type="string">
        If provided, this overrides the type preset applied to the text input label. Otherwise,
        label030 or label040 is used depending on the `size` prop.
      </Prop>
      <Prop name="spaceInline" type="SpacingPresetKeys">
        This is a spacing preset token which is applied as a margin on the right of
        the text input label. Use this to space inline content across the page.
      </Prop>
      <Prop name="spaceStack" type="SpacingPresetKeys">
        This is a spacing preset token which is applied as a margin on the bottom of
        the text input label. Use this to space stacking content down the page.
      </Prop>
    </Prop>
    <Prop name="assistiveText" type="object">
      <Prop name="stylePreset" type="string">
        If provided, this overrides the style preset applied to the assistive text.
      </Prop>
      <Prop name="typographyPreset" type="string">
        If provided, this overrides the type preset applied to the text input. Otherwise,
        label020 or label030 is used depending on the `size` prop.
      </Prop>
    </Prop>
    Refer to the defaults below for the object structure:
    <CodeFromDefaultPresets componentName="textInput" />
  </Prop>
</LegacyBlock>

## Playground

<Playground
  componentName="TextInput"
  newskitPath="/text-input"
  component={state => (
    <TextInput label="Label" {...state} />
  )}
  knobs={[
    {
      name: 'Size',
      propName: 'size',
      options: [
        {
          label: 'TextInputSize.Small',
          value: TextInputSize.Small,
        },
        {
          label: 'TextInputSize.Medium',
          value: TextInputSize.Medium,
          isDefault: true,
        },
        {
          label: 'TextInputSize.Large',
          value: TextInputSize.Large,
        },
      ],
    },
    {
      name: 'Style Preset',
      propName: 'overrides',
      options: [
        {
          label: 'Default',
          value: {
            label: {
              stylePreset: undefined
            },
            input: {
              stylePreset: undefined
            },
            assistiveText: {
              stylePreset: undefined
            },
          },
          isDefault: true,
        },
      ],
    }, 
    {
      name: 'Typography Preset',
      propName: 'overrides',
      options: [
        {
          label: 'Default',
          value: {
            label: {
              typographyPreset: undefined,
            },
            input: {
              typographyPreset: undefined,
            },
            assistiveText: {
              typographyPreset: undefined,
            },
          },
          isDefault: true,
        },
        {
          label: 'Diffierent presets',
          value: {
            label:{
              typographyPreset: 'label050',
            },
            assistiveText: {
              typographyPreset: 'label050',
            },
          },
        },
      ],
    },
    {
      name: 'Spacing Preset',
      propName: 'overrides',
      options: [
        {
          label: 'Default',
          value: {
            label:{
              spaceStack: undefined,
            },
            input: {
              spaceStack: undefined,
            },
          },
          isDefault: true,
        },
        {
          label: 'Diffierent presets',
          value: {
            input: {
              spaceStack: 'space040',
            },
          },
        },
      ],
    },
    {
      name: 'Assistive Text',
      propName: 'assistiveText',
      options: [
        {
          label: 'No Assistive Text',
          isDefault: true,
        },
        {
          label: 'With Assistive Text',
          value: 'Assistive text'
        },
      ],
    },
    {
      name: 'Placeholder',
      propName: 'placeholder',
      options: [
        {
          label: 'Default',
          isDefault: true,
        },
        {
          label: 'Custom Placeholder',
          value: 'This is some text'
        },
      ],
    },
    {
      name: 'Hide Label',
      propName: 'hideLabel',
      type: 'boolean',
      value: false
    },
    {
      name: 'Width',
      propName: 'overrides',
      options: [
        {
          label: 'Default',
          value: {
            width: undefined,
          },
          isDefault: true,
        },
        {
          label: 'Full-width',
          value: {
            width: '100%',
          }
        },
        {
          label: 'Fix-width',
          value: {
            width: '200px',
          },
        },
      ],
    },
    {
      name: 'Disabled',
      propName: 'disabled',
      type: 'boolean',
      value: false,
    },
    {
      name: 'Read-only',
      propName: 'readOnly',
      type: 'boolean',
      value: false,
    },
  ]}
/>
