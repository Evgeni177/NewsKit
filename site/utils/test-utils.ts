import {
  renderToFragmentWithThemeFactory,
  renderWithThemeFactory,
} from 'newskit/test/test-utils';
import {docsThemeLight} from '../theme/doc-theme';

export const renderWithTheme = renderWithThemeFactory(docsThemeLight);

export const renderToFragmentWithTheme = renderToFragmentWithThemeFactory(
  docsThemeLight,
);
