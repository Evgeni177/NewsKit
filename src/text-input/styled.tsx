import {
  styled,
  getTypographyPreset,
  getSpacingStack,
  getSpacingInline,
  getMinHeight,
  getPaddingPreset,
  getSizingFromTheme,
} from '../utils/style';
import {TextInputProps, TextInputSize} from './types';
import {TextBlock, TextBlockProps} from '../text-block/text-block';
import {getStylePreset, getStylePresetFromTheme} from '../utils/style-preset';
import {getToken} from '../utils/get-token';

export const StyledTextInputContainer = styled.div<TextInputProps>`
  ${({theme, overrides}) => {
    const getSizing = (tokenName: string) => {
      const token = getToken({theme, overrides}, '', '', tokenName);
      return getSizingFromTheme(token, undefined, true)({
        theme,
      });
    };
    const width = getSizing('width');
    return {
      width,
    };
  }}
`;

interface InputContainer
  extends Pick<TextInputProps, 'overrides' | 'disabled'> {
  $size: TextInputSize;
  stylePreset?: string;
  id?: string;
}

export const StyledInput = styled.input<InputContainer>`
  box-sizing: border-box;
  width: 100%;
  cursor: ${({disabled}) => (disabled ? 'not-allowed' : 'default')};
  ${({$size}) => getStylePreset(`textInput.${$size}.input`, 'input')}
  ${({$size}) =>
    getTypographyPreset(`textInput.${$size}.input`, 'input', {
      withCrop: true,
    })}
  ${({$size}) => getPaddingPreset(`textInput.${$size}.input`, 'input')}
  ${({$size}) => getSpacingStack(`textInput.${$size}.input`, 'input')}
  ${({$size}) => getSpacingInline(`textInput.${$size}.input`, 'input')} 
  min-height: ${({$size}) => getMinHeight(`textInput.${$size}.input`, 'input')} 
`;

export const StyledLabel = styled.label<InputContainer>`
    display: block;
    ${({$size}) => getStylePreset(`textInput.${$size}.label`, 'label')}
    ${({$size}) =>
      getTypographyPreset(`textInput.${$size}.label`, 'label', {
        withCrop: true,
      })}
    ${({$size}) => getSpacingStack(`textInput.${$size}.label`, 'label')}
    ${({$size}) => getSpacingInline(`textInput.${$size}.label`, 'label')};
 `;

export const StyledAssistiveText = styled(TextBlock)<
  Omit<TextBlockProps, 'overrides'> & InputContainer
>`
  ${({$size}) =>
    getStylePreset(`textInput.${$size}.assistiveText`, 'assistiveText')}
  ${({disabled}) =>
    getStylePresetFromTheme('textInputAssistiveText', 'stylePreset', {
      isDisabled: disabled,
    })}
  ${({$size}) =>
    getTypographyPreset(`textInput.${$size}.assistiveText`, 'assistiveText', {
      withCrop: true,
    })}
`;
