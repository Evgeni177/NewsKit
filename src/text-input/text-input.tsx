import React, {useEffect, useContext } from 'react';
import {useFormContext} from 'react-hook-form/dist/index.ie11';
import composeRefs from '@seznam/compose-react-refs'
import {TextInputProps, TextInputSize} from './types';
import {getSSRId} from '../utils/get-ssr-id';
import {FormValidationContext} from '../form/context'

import {
  StyledTextInputContainer,
  StyledInput,
  StyledAssistiveText,
  StyledLabel,
  InputIconContainer,
  IconContainer,
  StyledAssistiveTextContainer
} from './styled';
import { getToken } from '../utils/get-token';
import { useTheme } from '../theme';
import { IconFilledCheckCircle, IconFilledError } from '../icons';
import { FieldsHadErrorObject } from '../form/types';

export const TextInput = React.forwardRef<HTMLInputElement, TextInputProps>(({
  overrides = {},
  size = TextInputSize.Medium,
  label,
  disabled,
  hideLabel = false,
  ariaLabel,
  placeholder,
  assistiveText,
  rules,
  name,
  spellCheck = false,
  ...props
}, ref) => {

  const theme = useTheme();
  const {validationMode, setFieldsHadError, fieldsHadError} = useContext(FormValidationContext);
  const formContext = useFormContext();

  const hadError = name ? fieldsHadError[name] && fieldsHadError[name].hadError : undefined
  const errorText =
    formContext &&
    formContext.errors &&
    name &&
    formContext.errors[name] &&
    formContext.errors[name].message;

  useEffect(() => {
    if (!hadError && errorText && name) {
      const updateForFieldsHadError: FieldsHadErrorObject = {}

      updateForFieldsHadError[name] = {hadError: true}

      setFieldsHadError({
        ...fieldsHadError,
        ...updateForFieldsHadError
      })
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [errorText, hadError, name]);

  const id = getSSRId();
  const assistiveTextId = errorText && `${id}-error-text` || assistiveText && `${id}-assistive-text`

  const handleOnBlur = ({target: { value }}: {target: {value: string}}) => {
    if (validationMode === 'onBlur' && !hadError && name && (errorText || value)) {
        const updateForFieldsHadError: FieldsHadErrorObject = {}

        updateForFieldsHadError[name] = {hadError: true}
        setFieldsHadError(updateForFieldsHadError)
    }
  }

  const spaceInsetRight = getToken(
    {theme, overrides},
    `textInput.${size}.input`,
    'input',
    'spaceInsetRight',
  );

  const iconSpace = getToken(
    {theme, overrides},
    `textInput.${size}.input`,
    'input',
    'iconSpace',
  );

  const iconSize = getToken(
    {theme, overrides},
    `textInput.${size}.input`,
    'input',
    'iconSize',
  );

  const valid = formContext?.formState?.isSubmitSuccessful || (hadError && !errorText);

  return (
    <StyledTextInputContainer label={label} overrides={overrides}>
      <StyledLabel $size={size} htmlFor={id} overrides={overrides}>
        {!hideLabel && label}
      </StyledLabel>
      <InputIconContainer>
        <StyledInput
          ref={composeRefs(formContext?.register(rules), ref)}
          name={name}
          type="text"
          placeholder={placeholder}
          id={id}
          aria-describedby={assistiveTextId}
          disabled={disabled}
          aria-label={hideLabel === true ? label : ariaLabel}
          $size={size}
          {...props}
          overrides={overrides}
          invalid={!!errorText}
          valid={valid}
          aria-invalid={!!errorText}
          onBlur={handleOnBlur}
          spaceInsetRight={spaceInsetRight}
          spellCheck={spellCheck}
        />
        {(!!errorText || valid) && (
          <IconContainer iconSpace={iconSpace}>
            {
              !!errorText && <IconFilledError data-testid='error-icon' overrides={{ size: iconSize, stylePreset: 'iconNegative' }}  /> ||
              valid && <IconFilledCheckCircle data-testid='tick-icon' overrides={{ size: iconSize, stylePreset: 'iconPositive'}} />
            }
          </IconContainer>
        )}
      </InputIconContainer>
      <StyledAssistiveTextContainer $size={size}>
        {((assistiveText || errorText) && !valid) && (
          <StyledAssistiveText
            $size={size}
            id={assistiveTextId}
            disabled={disabled}
            overrides={overrides}
            invalid={!!errorText}
            valid={valid}
            role={errorText && 'alert'}
            aria-live={errorText && 'polite'}
          >
            {errorText || assistiveText}
          </StyledAssistiveText>
        )}
      </StyledAssistiveTextContainer>
    </StyledTextInputContainer>
  );
});

TextInput.displayName = 'TextInput';
