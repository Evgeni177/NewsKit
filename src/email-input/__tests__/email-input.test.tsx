import React from 'react';
import {EmailInput} from '..';
import {renderToFragmentWithTheme} from '../../test/test-utils';
import {TextInputProps} from '../../text-input';

const renderEmailInputWithLabelHidden = (props: TextInputProps) => (
  <EmailInput hideLabel assistiveText="assistiveText" {...props} />
);

describe('EmailInput', () => {
  test('renders correctly with label hidden', () => {
    const fragment = renderToFragmentWithTheme(renderEmailInputWithLabelHidden);
    expect(fragment).toMatchSnapshot();
  });
});
