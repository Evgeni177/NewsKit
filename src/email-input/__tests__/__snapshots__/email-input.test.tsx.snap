// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmailInput renders correctly with label hidden 1`] = `
<DocumentFragment>
  .emotion-1 {
  display: block;
  color: #0a0a0a;
  font-family: "Noto Sans",sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 500;
  letter-spacing: 0;
  margin-bottom: 16px;
  margin-right: 4px;
}

.emotion-1 svg {
  fill: #0a0a0a;
}

.emotion-1::before {
  content: '';
  display: block;
  height: 0;
  width: 0;
  margin-bottom: -0.36250000000000004em;
}

.emotion-1::after {
  content: '';
  display: block;
  height: 0;
  width: 0;
  margin-top: -0.4625em;
}

.emotion-2 {
  position: relative;
}

.emotion-3 {
  box-sizing: border-box;
  width: 100%;
  cursor: default;
  background-color: transparent;
  border-style: solid;
  border-color: #e4e4e4;
  border-width: 1px;
  border-radius: 8px;
  color: #2e2e2e;
  text-overflow: ellipsis;
  font-family: "Noto Sans",sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 400;
  letter-spacing: 0;
  padding: 12px;
  margin-bottom: 8px;
  min-height: 48px;
}

.emotion-3::-webkit-input-placeholder {
  color: #535353;
}

.emotion-3::-moz-placeholder {
  color: #535353;
}

.emotion-3:-ms-input-placeholder {
  color: #535353;
}

.emotion-3::placeholder {
  color: #535353;
}

.emotion-3:focus:not(:disabled) {
  background-color: #fafafa;
  border-color: #0a68c1;
}

.emotion-3:hover:not(:disabled) {
  background-color: #fafafa;
}

.emotion-3:disabled {
  border-color: #e4e4e4;
  color: #c0c0c0;
}

.emotion-3::before {
  content: '';
  display: block;
  height: 0;
  width: 0;
  margin-bottom: -0.36250000000000004em;
}

.emotion-3::after {
  content: '';
  display: block;
  height: 0;
  width: 0;
  margin-top: -0.4625em;
}

.emotion-4 {
  min-height: 12px;
}

.emotion-5 {
  margin: 0;
  padding: 1px 0;
  color: #535353;
  font-family: "Noto Sans",sans-serif;
  font-size: 14px;
  line-height: 1.4285714285714286;
  font-weight: 500;
  letter-spacing: 0;
}

.emotion-5:disabled {
  color: #c0c0c0;
}

.emotion-5::before {
  content: '';
  display: block;
  height: 0;
  width: 0;
  margin-bottom: -0.32678571428571435em;
}

.emotion-5::after {
  content: '';
  display: block;
  height: 0;
  width: 0;
  margin-top: -0.4267857142857143em;
}

<div
    class="emotion-0"
  >
    <label
      class="emotion-1"
      for="mock-nk-1"
    />
    <div
      class="emotion-2"
    >
      <input
        aria-describedby="mock-nk-1-assistive-text"
        aria-invalid="false"
        class="emotion-3"
        id="mock-nk-1"
        spellcheck="false"
        type="email"
      />
    </div>
    <div
      class="emotion-4"
    >
      <p
        class="emotion-5"
        id="mock-nk-1-assistive-text"
      >
        assistiveText
      </p>
    </div>
  </div>
</DocumentFragment>
`;
